dwave/optimization/model.pyx

```diff
@@ -1764,7 +1764,24 @@ cdef class ArraySymbol(Symbol):
         if exponent == 2:
             from dwave.optimization.symbols import Square  # avoid circular import
             return Square(self)
-        raise ValueError("only squaring is currently supported")
+        # check if exponent is an integer greater than 2
+        elif (isinstance(exponent, int) and exponent > 2) or (isinstance(exponent, float) and exponent.is_integer() and exponent > 2):
+            from dwave.optimization.symbols import Square  # avoid circular import
+            from dwave.optimization.symbols import Multiply  # avoid circular import
+            nodes = [self]
+            nodes_exponents = [1]
+            while nodes_exponents[-1] < exponent:
+                if nodes_exponents[-1] * 2 <= exponent:
+                    nodes.append(Square(nodes[-1]))
+                    nodes_exponents.append(nodes_exponents[-1] * 2)
+                else:
+                    indexer = 2
+                    while nodes_exponents[-1] + nodes_exponents[-indexer] > exponent:
+                        indexer += 1
+                    nodes.append(Multiply(nodes[-indexer], nodes[-1]))
+                    nodes_exponents.append(nodes_exponents[-indexer] + nodes_exponents[-1])
+            return nodes[-1]
```

Comment:
We can take advantage of some other operators
```Python
            expanded = itertools.repeat(self, int(exponent))
            out = next(expanded)  # get the first one
            for symbol in expanded:
                out *= symbol  # multiply it by the remainder
            return out
```

Reply:
Understood. Thank you!

Reply:
I added this change into the commit bd3a1ae0eff222ae43c548480dc28521b23762e6. However this appears to be causing the tests to fail. I haven't been able to figure out why yet.

Reply:
@jwoo098, your previous implementation might be more verbose, but it creates a lot more efficient encoding than this naive multiplication.

Something like $O(\log_2^2{n})$ `Square` + $O(\log_2{n})$ `Multiply` operations, vs $O(n)$ `Multiply` operations for the naive implementation.

If anything, I'd suggest you update it to be more efficient, rather than less. By using basic method of [exponentiation by squaring](https://en.wikipedia.org/wiki/Exponentiation_by_squaring), you should be able to get $O(\log_2{n})$ `Square` + $O(\log_2{n})$ `Multiply` ops.

Reply:
And btw, the test is failing because it's [expecting `Square` symbol](https://github.com/dwavesystems/dwave-optimization/blob/0578cc0e7aba80f6255e3b77f94dcf5dd5197e26/tests/test_model.py#L75) as a result of `x**2`, and it's getting `NaryMultiply`.

Reply:
> @jwoo098, your previous implementation might be more verbose, but it creates a lot more efficient encoding than this naive multiplication.
> 
> Something like O ( log 2 2 ⁡ n ) `Square` + O ( log 2 ⁡ n ) `Multiply` operations, vs O ( n ) `Multiply` operations for the naive implementation.
> 
> If anything, I'd suggest you update it to be more efficient, rather than less. By using basic method of [exponentiation by squaring](https://en.wikipedia.org/wiki/Exponentiation_by_squaring), you should be able to get O ( log 2 ⁡ n ) `Square` + O ( log 2 ⁡ n ) `Multiply` ops.

@randomir 
When using *= or __imul__(), it appears that I only get one additional node from __pow__ no matter how high the exponent becomes. However, when I use Multiply() (as in the verbose implementation) I get the ceiling of log2(exponent)-1 additional nodes in my more verbose implementation of __pow__ using Multiply(). Is the verbose implementation still better despite the higher number of nodes created in the model?

Reply:
You're getting only one node, `NaryMultiply` that multiples the input $n-1$ times.

IMO, the best option would be to implement `Pow` symbol in C++, with the underlying optimized exponentiation.

Second best would be exponentiation by squaring (or similar method) in Python, with $O(\log_2{n})$ nodes created (these are essentially multiply nodes, square being a special case).

And probably the least efficient is $n-1$ multiplications in `NaryMultiply`. But that one has the benefit of being the simplest, most straightforward implementation.

My assumptions about runtime performance of additional $O(\log_2{n})$ (multiply) nodes vs a single $n-1$ multiply node might be wrong, though, as I haven't done such benchmarks. I.e. I'm not sure about the order of constant factors that come into play here. @arcondello should weight in.

Reply:
Yes, the best option would be to implement a C++ `Pow` symbol.

In the mean time, it's generally more efficient to have fewer "large" symbols than many "smaller" symbols. So my expectation is that unless the exponent is truly large (which would lead to many other problems), that having a single node will be faster in practice.

Also, because we intend to replace with with a `Pow` symbol in the near future, having a single node and a simple implementation makes that transition easier. Users won't suddenly see a change in the number of nodes etc etc.

For those reasons, I'd like to use the `NaryMultiply` node approach as I suggested above.

Reply:
In that case, the simplicity of your suggested approach makes perfect sense!